File: nativeLib/rary.ohd.pproc/src/GageQCEngine/TEXT/check_consistency.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "gageqc_defs.h"
#include "gageqc_types.h"

void check_consistency(int j,
		       struct station * station,
		       int max_stations)

{

extern int mpe_dqc_max_precip_neighbors;
int k,m;
float rtotal;
extern struct pdata pdata[10];
int i,ii;

for(k=0;k<max_stations;k++) {

        pdata[j].stn[k].tcons=1;
	 
	if(pdata[j].stn[k].frain[4].data < 0 ||
	   pdata[j].stn[k].frain[4].qual==1)
	                                 continue;
	
        rtotal=0;
        for(m=0;m<4;m++) {
	 
                if(pdata[j].stn[k].frain[m].data >= 0)
                            rtotal=rtotal+pdata[j].stn[k].frain[m].data;
        
	         }
	      
        if(fabs(rtotal-pdata[j].stn[k].frain[4].data) > .01) 
	                pdata[j].stn[k].tcons=-1;
			
			
  }
 
for(k=0;k<max_stations;k++) {

        for(m=0;m<5;m++) {

            if(pdata[j].stn[k].frain[m].qual==1 ||
	       pdata[j].stn[k].frain[m].data < 0)
	                           continue;

             pdata[j].stn[k].scons[m]=1;
	     
	     for(ii=0; ii < mpe_dqc_max_precip_neighbors; ii++){
   
                 i=station[k].index[ii];

                 if(station[i].lat == station[k].lat &&
                    station[i].lon == station[k].lon &&
		    pdata[j].stn[i].frain[m].qual != 1 &&
		    pdata[j].stn[i].frain[m].data >=0 &&
		    pdata[j].stn[i].frain[m].data !=
		    pdata[j].stn[k].frain[m].data) {
		    
		       pdata[j].stn[k].scons[m]=-1;
		       
		       }
		         
		    }
		    
		    
		}    
		    
  }
	       
      			

/*  ==============  Statements containing RCS keywords:  */
{static char rcs_id1[] = "$Source$";
 static char rcs_id2[] = "$Id$";}
/*  ===================================================  */

}			
			
                  

File: cave/com.raytheon.viz.mpe/src/com/raytheon/viz/mpe/util/CheckConsistency.java
/**
 * This software was developed and / or modified by Raytheon Company,
 * pursuant to Contract DG133W-05-CQ-1067 with the US Government.
 * 
 * U.S. EXPORT CONTROLLED TECHNICAL DATA
 * This software product contains export-restricted data whose
 * export/transfer/disclosure is restricted by U.S. law. Dissemination
 * to non-U.S. persons whether in the United States or abroad requires
 * an export license or other authorization.
 * 
 * Contractor Name:        Raytheon Company
 * Contractor Address:     6825 Pine Street, Suite 340
 *                         Mail Stop B8
 *                         Omaha, NE 68106
 *                         402.291.0100
 * 
 * See the AWIPS II Master Rights File ("Master Rights File.pdf") for
 * further licensing information.
 **/
package com.raytheon.viz.mpe.util;

import java.util.List;

import com.raytheon.viz.mpe.util.DailyQcUtils.Station;

/**
 * Checks consistency of data.
 * 
 * <pre>
 * 
 * SOFTWARE HISTORY
 * Date         Ticket#    Engineer    Description
 * ------------ ---------- ----------- --------------------------
 * Mar 10, 2009            snaples     Initial creation
 * May 21, 2018   7131     mduff       Updated to accommodate changes to other classes and cleanup.
 * 
 * </pre>
 * 
 * @author snaples
 */

public class CheckConsistency {

    public void check_consistency(int j, List<Station> precip_stations,
            int numPstations) {

        int dqc_neig = DailyQcUtils.mpe_dqc_max_precip_neighbors;
        int max_stations = numPstations;
        float rtotal;

        for (int k = 0; k < max_stations; k++) {
            DailyQcUtils.pdata[j].stn[k].tcons = 1;

            if (DailyQcUtils.pdata[j].stn[k].frain[4].data < 0) {
                continue;
            }

            rtotal = 0;
            for (int m = 0; m < 4; m++) {
                if (DailyQcUtils.pdata[j].stn[k].frain[m].data >= 0) {
                    rtotal += DailyQcUtils.pdata[j].stn[k].frain[m].data;
                }

            }

            if (Math.abs(rtotal
                    - DailyQcUtils.pdata[j].stn[k].frain[4].data) > .01) {
                DailyQcUtils.pdata[j].stn[k].tcons = -1;
            }
        }

        for (int k = 0; k < max_stations; k++) {
            for (int m = 0; m < 5; m++) {
                if (DailyQcUtils.pdata[j].stn[k].frain[m].qual == 1
                        || DailyQcUtils.pdata[j].stn[k].frain[m].data < 0) {
                    continue;
                }

                DailyQcUtils.pdata[j].stn[k].scons[m] = 1;
                for (int ii = 0; ii < dqc_neig; ii++) {
                    int i = precip_stations.get(k).index[ii];

                    if ((precip_stations.get(i).lat == precip_stations
                            .get(k).lat)
                            && (precip_stations.get(i).lon == precip_stations
                                    .get(k).lon)
                            && (DailyQcUtils.pdata[j].stn[i].frain[m].qual != 1)
                            && (DailyQcUtils.pdata[j].stn[i].frain[m].data >= 0)
                            && (DailyQcUtils.pdata[j].stn[i].frain[m].data != DailyQcUtils.pdata[j].stn[k].frain[m].data)) {

                        DailyQcUtils.pdata[j].stn[k].scons[m] = -1;
                    }
                }
            }
        }
    }
}
